# 01. Importing libraries

In [2]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 2. Paths

In [4]:
path = r'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

In [5]:
path

'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

# 03. Importing Data

In [10]:
#Import ords_prods_merge_agg' Dataset\n",
ords_prods_merge = pd.read_pickle(os.path.join(path, '02-Data', 'Prepared-Data', 'ords_prods_merge_agg.pkl'))

In [11]:
ords_prods_merge.shape

(32404859, 24)

In [12]:
#Import 'customers' Dataset
df_cust = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'customers.csv'))

In [13]:
df_cust.shape

(206209, 10)

# 04. Data Wrangling

In [19]:
# Checking the dataframe
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [21]:
#Printing the Stats of the 'df_cust' Dataset
df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [23]:
# Renaming the n_dependants varaible to number_dependants in the 'df_cust' Dataset
df_cust.rename(columns = {'n_dependants' : 'number_dependants'}, inplace = True)

In [25]:
#Checking the 'df_cust' Dataset
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374

In [27]:
#Dropping the columns 'First Name' &  'Surnam'	and assinged it to a new datafarame named 'df_customers'
df_customers = df_cust.drop(columns = ['First Name', 'Surnam'])

In [29]:
#Checking the 'df_customers' Dataset
df_customers.head()

user_id  Gender       STATE  Age date_joined  number_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017                  3    married   
1    33890  Female  New Mexico   36    1/1/2017                  0     single   
2    65803    Male       Idaho   35    1/1/2017                  2    married   
3   125935  Female        Iowa   40    1/1/2017                  0     single   
4   130797  Female    Maryland   26    1/1/2017                  1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

In [31]:
df_customers.shape

(206209, 8)

In [33]:
#Checking the Columns Data Types in the 'df_customers' Dataset
df_customers.dtypes

user_id               int64
Gender               object
STATE                object
Age                   int64
date_joined          object
number_dependants     int64
fam_status           object
income                int64
dtype: object

In [35]:
#Changing the data type of 'user_id' variable to string in the 'df_customers' dataframe
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [37]:
#Checking the datatype of the 'user_id' column in 'df_customers'dataframe
df_customers['user_id'].dtype

dtype('O')

In [69]:
#Checking the datatype of the 'user_id' column in the 'ords_prods_merge' dataframe
ords_prods_merge['user_id'].dtype

dtype('int64')

In [71]:
#Changing the data type of 'user_id' variable to string in the 'ords_prods_merge' dataframe
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [73]:
#Checking the datatype of the 'user_id' column in the 'ords_prods_merge' dataframe
ords_prods_merge['user_id'].dtype

dtype('O')

In [75]:
#Checking the columns of the 'ords_prods_merge' column 
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id user_id  order_number  orders_day_of_week  order_hour_of_day  ...  \
0   3139998     138            28                   6                 11  ...   
1   1977647     138            30                   6                 17  ...   
2    389851     709             2                   0                 21  ...   
3    652770     764             1                   3                 13  ...   
4   1813452     764             3                   4                 17  ...   

     price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days           Most orders   
1  Mid-range product  Regularly busy  Regular days        Average orders   
2  Mid-range product     Busiest day  Busiest days        Average orders   
3  Mid-range product  Regularly busy  Slowest days           Most orders   
4  Mid-range product      Least busy  Slowest days        Average orders   

  max_order      loyalty_flag Avg_spending spending_flag  \
0        32  Regular customer     6.935811   Low spender   
1        32  Regular customer     6.935811   Low spender   
2         5      New customer     7.930208   Low spender   
3         3      New customer     4.972414   Low spender   
4         3      New customer     4.972414   Low spender   

   median_days_since_last_order order_frquency_flag  
0                           8.0   Frequent customer  
1                           8.0   Frequent customer  
2                           8.0   Frequent customer  
3                           9.0   Frequent customer  
4                           9.0   Frequent customer  

[5 rows x 24 columns]

In [77]:
#Checking the count/frequency of the 'Gender' variable
df_customers['Gender'].value_counts()

Gender
Male      104067
Female    102142
Name: count, dtype: int64

In [79]:
#Checking the  The count/frequency of the 'fam_status' variable
df_customers['fam_status'].value_counts()

fam_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [81]:
#Checking the  count/frequency of the 'STATE' variable
df_customers['STATE'].value_counts()

STATE
Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana 

In [83]:
#Finding how many 'STATE' there is
df_customers['STATE'].nunique()

51

In [85]:
#Check if the 'df_customers' dataset is a dartaframe 
type(df_customers) == pd.DataFrame

True

In [87]:
#Finding full duplicates within the 'df_customers' dataframe
df_customers[df_customers.duplicated()]

Empty DataFrame
Columns: [user_id, Gender, STATE, Age, date_joined, number_dependants, fam_status, income]
Index: []

In [89]:
#Find missing values within the 'df_customers' dataframe
df_customers.isnull().sum()

user_id              0
Gender               0
STATE                0
Age                  0
date_joined          0
number_dependants    0
fam_status           0
income               0
dtype: int64

In [91]:
#Find missing values within the 'df_cust' dataframe
df_cust.isnull().sum()

user_id                  0
First Name           11259
Surnam                   0
Gender                   0
STATE                    0
Age                      0
date_joined              0
number_dependants        0
fam_status               0
income                   0
dtype: int64

In [93]:
#Creating fuction to search for a mixed data type in a column
for col in df_customers.isnull().sum().columns.tolist():
  weird = (df_customers.isnull().sum()[[col]].applymap(type) != df_customers.isnull().sum()[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers.isnull().sum()[weird]) > 0:
    print (col)

AttributeError: 'Series' object has no attribute 'columns'

In [95]:
#Checking the unique 'user_id' unique 'user_id' in the 'ords_prods_merge' dataframe
ords_prods_merge.groupby('user_id').size()

user_id
1          59
10        143
100        27
1000      103
10000    1092
         ... 
99995      50
99996     128
99997      36
99998      21
99999     329
Length: 206209, dtype: int64

In [97]:
#Checking the number of unique 'user_id' in the 'df_customers' dataframe
df_customers.groupby('user_id').size()

user_id
1        1
10       1
100      1
1000     1
10000    1
        ..
99995    1
99996    1
99997    1
99998    1
99999    1
Length: 206209, dtype: int64

In [109]:
#Combining the 'df_customers' dataframe with the 'ords_prods_merge' dataframe using an inner join on the 'user_id' column, and 
#assigning the result to a new dataframe called 'ords_prods_customers_merge'
ords_prods_customers_merge = ords_prods_merge.merge(df_customers, on = ['user_id'], how = 'inner')


In [115]:
#Checking the shape of the dataframe
ords_prods_customers_merge.shape

(32404859, 31)

In [117]:
ords_prods_customers_merge['user_id'].value_counts()

user_id
201268    3704
129928    3637
164055    3061
186704    2936
176478    2921
          ... 
195256       3
55712        3
203875       2
124615       2
91567        1
Name: count, Length: 206209, dtype: int64

In [119]:
#Checking for missing values in the 'ords_prods_customers_merge' dataframe
ords_prods_customers_merge.isnull().sum()

product_id                            0
product_name                          0
aisle_id                              0
department_id                         0
prices                                0
order_id                              0
user_id                               0
order_number                          0
orders_day_of_week                    0
order_hour_of_day                     0
days_since_prior_order          2076096
add_to_cart_order                     0
reordered                             0
_merge                                0
price_range_loc                       0
busiest_day                           0
Busiest_days                          0
busiest_period_of_day                 0
max_order                             0
loyalty_flag                          0
Avg_spending                          0
spending_flag                         0
median_days_since_last_order          5
order_frquency_flag                   5
Gender                                0


In [121]:
#Checking foo full duplicates within the 'ords_prods_customers_merge' dataframe
ords_prods_customers_merge[ords_prods_customers_merge.duplicated()]

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, add_to_cart_order, reordered, _merge, price_range_loc, busiest_day, Busiest_days, busiest_period_of_day, max_order, loyalty_flag, Avg_spending, spending_flag, median_days_since_last_order, order_frquency_flag, Gender, STATE, Age, date_joined, number_dependants, fam_status, income]
Index: []

[0 rows x 31 columns]

# Observations:

### The new dataframe 'df_customers' contains 206209 columns x 10 rows

## Summary statistics:
The summary statistics make sense--nothing appears to be out of place. 
The 'Age' range is from 18 years to 81 years, with a mean and median of 49 years. 
In terms of income, the range is from $25,903 to $593,902, with a median of $93,547 and a mean of $94,632.
The maximum value for all entries is consistent with the total number of data observations/rows, which is 206209.

## Rename Columns:
The data variables are clear and do not require any renaming of columns. However, if needed, the column 'n_dependants' could be changed to 'number_dependants' for better clarity.
The column 'n_dependants' changed to 'number_of_dependants'

## Drop Columns:
The 'First name' and 'Surname' columns are unnecessary for the analysis and can be dropped. 
created a new dataframe named 'df_customers', these columns were removed, resulting in a dataframe of 206209 rows and 8 columns.

## Data Type:
#### The datatype of the 'user_id' is int64. 
#### The datatype of the 'user_id' is changed to string. 
#### No other datatype changes are needed.

## Checking the count of the 'Gender' column:
#### There are more male customers than female customers in Instacart

## Checking the count of the 'fam_status' columns:
#### There are more 'married' customers than customers in any other category in Instacart.

## Checking the count of the 'STATE' columns:
#### There are 51 states (including DC).

## Duplicates:
#### There are no duplicates in the dataframe.

## Missing values:
#### There are no missing values in the dataframe 'df_customers'

#### Just for curiosity, I checked for missing values in the 'df_customers' table where the 'First Name' and 'Surname' columns are present. I found 11259 missing values in the 'First Name' column

## Mixed datatype:
#### There are no mixed datatypes

#### Changing the datatype of the "user_id" column in the "ords_prods_merge" from "int64" to "string" datatype.

## Combining Dataframes:
#### Combining the 'df_customers' dataframe with 'ords_prods_merge' dataframe using an inner join on the 'user_id' column, as we want to keep the data from both dataframes.

#### The new dataframe 'ords_prods_customers_merge' contains 32,404,859 rows and 31 columns, with a total of 206,209 unique "user_id". 

#### The dataframe, 'ords_prods_customers_merge', contains missing values for the column 'days_since_prior_order', which is expected as it denotes the very first order each customer placed.

####  There are a total of 5 missing values found in each of the columns 'median_days_since_last_order' and 'order_frequency_flag'.
####  The dataframe, 'ords_prods_customers_merge', contains no duplicates.

# 05. Export the dataframe

In [130]:
# Export the 'ords_prods_customers_merge' dataframe to pkl formate
ords_prods_customers_merge.to_pickle(os.path.join(path, '02-Data','Prepared-Data', 'ords_prods_customers_merge.pkl'))